update(JS): web/javascript/reference/iteration_protocols (#3132)
webdoky/content: files/uk/web/javascript/reference/iteration_protocols/index.md

### Погано сформовані ітеровані об'єкти

Помилки можуть відбуватися під час отримання ітератора з ітерованого об'єкта. Мовний інваріант, накладений тут, – це те, що ітерований об'єкт повинен видавати чинний ітератор:

- Він має викличний метод `[Symbol.iterator]()`.
- Метод `[Symbol.iterator]()` повертає об'єкт.
- Об'єкт, повернений `[Symbol.iterator]()`, має викличний метод `next()`.

Коли для ініціації ітерування на погано сформованому ітерованому об'єкті вживається вбудований синтаксис, викидається TypeError.

```js example-bad
const nonWellFormedIterable = { [Symbol.iterator]: 1 };
[...nonWellFormedIterable]; // TypeError: nonWellFormedIterable is not iterable
nonWellFormedIterable[Symbol.iterator] = () => 1;
[...nonWellFormedIterable]; // TypeError: [Symbol.iterator]() returned a non-object value
nonWellFormedIterable[Symbol.iterator] = () => ({});
[...nonWellFormedIterable]; // TypeError: nonWellFormedIterable[Symbol.iterator]().next is not a function
```

> У разі асинхронних ітерованих, то якщо їхня властивість `[Symbol.asyncIterator]()` має значення `undefined` або `null`, то JavaScript відступає до вживання натомість властивості `[Symbol.iterator]` (і загортає результівний ітератор у асинхронний ітератор, [переспрямовуючи](#perespriamuvannia-pomylok) методи). Інакше – властивість `[Symbol.asyncIterator]` також повинна відповідати інваріантам вище.

Author: ```suggestion
У разі асинхронних ітерованих, то якщо їхня властивість `[Symbol.asyncIterator]()` має значення `undefined` або `null`, то JavaScript відступає до вживання натомість властивості `[Symbol.iterator]` (і загортає результівний ітератор в асинхронний ітератор, [переспрямовуючи](#perespriamuvannia-pomylok) методи). Інакше – властивість `[Symbol.asyncIterator]` також повинна відповідати інваріантам вище.
```

Такого роду помилкам можна запобігти, валідуючи ітерований об'єкт перед спробою його ітерування. Проте таке трапляється доволі рідко, адже зазвичай тип об'єкта, що ітерується, відомий. Якщо ви отримуєте ітерований об'єкт з якогось іншого коду, то вам слід просто дати помилці поширитися до викликача, щоб він знав: було надано невалідні вихідні дані.

### Помилки під час ітерування

Більшість помилок трапляються під час обходження ітератора (викликаючи `next()`). Мовний інваріант, що тут накладається, полягає в тому, що метод `next()` повинен повертати об'єкт (у разі асинхронних ітераторів – об'єкт після очікування). Інакше - викидається TypeError.

Якщо інваріант порушено або метод `next()` викидає помилку (для асинхронних ітераторів також може бути повернення відхиленого промісу), то помилка поширюється до викликача. У вбудованих записах синтаксису така ітерація переривається без повторних спроб і очищення (виходячи з припущення, що якщо метод `next()` викинув помилку, то все уже очищено). Якщо ви вручну викликаєте `next()`, то можете перехопити помилку і спробувати повторно викликати `next()`, проте в загальному слід припускати, що ітератор вже закрито.

> Якщо викликач вирішить вийти з ітерування з якихось причин, відмінних від помилок, описаних у абзаці вище, наприклад, коли він перейшов до стану помилки у власному коді (наприклад, у разі обробки невалідного значення, виданого ітератором), йому слід викликати метод `return()` на ітераторі, якщо такий є. Це дасть ітератору змогу виконати якесь очищення. Метод `return()` викликається лише для раннього виходу: якщо `next()` повертає `done: true`, то метод `return()` не викликається, виходячи з припущення, що ітератор уже очистився.

Author: ```suggestion
Якщо викликач вирішить вийти з ітерування з якихось причин, відмінних від помилок, описаних в абзаці вище, наприклад, коли він перейшов до стану помилки у власному коді (наприклад, у разі обробки невалідного значення, виданого ітератором), йому слід викликати метод `return()` на ітераторі, якщо такий є. Це дасть ітератору змогу виконати якесь очищення. Метод `return()` викликається лише для раннього виходу: якщо `next()` повертає `done: true`, то метод `return()` не викликається, виходячи з припущення, що ітератор уже очистився.
```

Метод `return()` так само може бути невалідним! Мова також вимагає, щоб метод `return()` повертав об'єкт, а інакше - викидав TypeError. Якщо метод `return()` викидає помилку, то ця помилка поширюється до викликача. Проте якщо метод `return()` викликається, тому що викликач зіткнувся з помилкою у своєму власному коді, то така помилка замінює помилку, викинуту методом `return()`.

Зазвичай викликач реалізує обробку помилок ось так:

```js
try {
  for (const value of iterable) {
    // ...
  }
} catch (e) {
  // Обробити помилку
}
```

> `catch` зможе перехопити помилки, викинуті, бо `iterable` не є валідним ітерованим об'єктом, бо `next()` викидає помилку, бо `return()` викидає помилку (якщо цикл `for` завершується передчасно) і бо тіло циклу `for` викидає помилку.

Author: ```suggestion
`catch` зможе перехопити помилки, викинуті, коли `iterable` не є валідним ітерованим об'єктом, коли `next()` викидає помилку, коли `return()` викидає помилку (якщо цикл `for` завершується передчасно) або коли тіло циклу `for` викидає помилку.
```

Більшість ітераторів реалізовані за допомогою функцій-генераторів, тож ми продемонструємо, як функції-генератори зазвичай обробляють помилки:

```js
function* gen() {
  try {
    yield doSomething();
    yield doSomethingElse();
  } finally {
    cleanup();
  }
}
```

Відсутність `catch` змушує помилки, викинуті `doSomething()` або `doSomethingElse()`, поширюватися до викликача `gen`. Якщо ці помилки перехоплюються всередині функції-генератора (що так само вітається), то така функція-генератор може вирішити продовжити видавати значення або завершитись передчасно. Проте блок `finally` необхідний для генераторів, що зберігають ресурси відкритими. Блок `finally` гарантовано запуститься, або коли востаннє викличеться `next()`, або коли викличеться `return()`.

### Переспрямування помилок

> Частина вбудованих записів синтаксису загортають ітератор у інший ітератор. Серед них ітератор, що створюється {{jsxref("Iterator.from()")}}, [ітераторні помічники](/uk/docs/Web/JavaScript/Reference/Global_Objects/Iterator#iteratorni-pomichnyky) (`map()`, `filter()`, `take()`, `drop()` і `flatMap()`), [`yield *`](/uk/docs/Web/JavaScript/Reference/Operators/yield*) і прихована обгортка, коли щодо синхронного ітератора використовується асинхронна ітерація (`for await...of`, `Array.fromAsync`). Загорнутий ітератор тоді відповідає за переспрямування помилок між внутрішнім ітератором і викликачем.

Author: ```suggestion
Частина вбудованих записів синтаксису загортають ітератор в інший ітератор. Серед них ітератор, що створюється {{jsxref("Iterator.from()")}}, [ітераторні помічники](/uk/docs/Web/JavaScript/Reference/Global_Objects/Iterator#iteratorni-pomichnyky) (`map()`, `filter()`, `take()`, `drop()` і `flatMap()`), [`yield *`](/uk/docs/Web/JavaScript/Reference/Operators/yield*) і прихована обгортка, коли щодо синхронного ітератора використовується асинхронна ітерація (`for await...of`, `Array.fromAsync`). Загорнутий ітератор тоді відповідає за переспрямування помилок між внутрішнім ітератором і викликачем.
```

[...nonWellFormedIterable]; // TypeError: nonWellFormedIterable[Symbol.iterator]().next is not a function
```

У разі асинхронних ітерованих, то якщо їхня властивість `[Symbol.asyncIterator]()` має значення `undefined` або `null`, то JavaScript відступає до вживання натомість властивості `[Symbol.iterator]` (і загортає результівний ітератор в асинхронний ітератор, [переспрямовуючи](#perespriamuvannia-pomylok) методи). Інакше – властивість `[Symbol.asyncIterator]` також повинна відповідати інваріантам вище.

Такого роду помилкам можна запобігти, валідуючи ітерований об'єкт перед спробою його ітерування. Проте таке трапляється доволі рідко, адже зазвичай тип об'єкта, що ітерується, відомий. Якщо ви отримуєте ітерований об'єкт з якогось іншого коду, то вам слід просто дати помилці поширитися до викликача, щоб він знав: було надано невалідні вихідні дані.

### Помилки під час ітерування

Більшість помилок трапляються під час обходження ітератора (викликаючи `next()`). Мовний інваріант, що тут накладається, полягає в тому, що метод `next()` повинен повертати об'єкт (у разі асинхронних ітераторів – об'єкт після очікування). Інакше - викидається TypeError.

> Якщо інваріант порушено або метод `next()` викидає помилку (для асинхронних ітераторів також може бути повернення відхиленого промісу), то помилка поширюється до викликача. У вбудованих записах синтаксису така ітерація переривається без повторних спроб і очищення (виходячи з припущення, що якщо метод `next()` викинув помилку, то все уже очищено). Якщо ви вручну викликаєте `next()`, то можете перехопити помилку і спробувати повторно викликати `next()`, проте в загальному слід припускати, що ітератор вже закрито.

Reviewer: ```suggestion
Якщо інваріант порушено або метод `next()` викидає помилку (для асинхронних ітераторів також може бути повернення відхиленого промісу), то помилка поширюється до викликача. У вбудованих записах синтаксису така ітерація переривається без повторних спроб і очищення (виходячи з припущення, що якщо метод `next()` викинув помилку, то все уже очищено). Якщо ви вручну викликаєте `next()`, то можете перехопити помилку і спробувати повторно викликати `next()`, проте загалом слід припускати, що ітератор вже закрито.
```

Якщо викликач вирішить вийти з ітерування з якихось причин, відмінних від помилок, описаних в абзаці вище, наприклад, коли він перейшов до стану помилки у власному коді (наприклад, у разі обробки невалідного значення, виданого ітератором), йому слід викликати метод `return()` на ітераторі, якщо такий є. Це дасть ітератору змогу виконати якесь очищення. Метод `return()` викликається лише для раннього виходу: якщо `next()` повертає `done: true`, то метод `return()` не викликається, виходячи з припущення, що ітератор уже очистився.

> Метод `return()` так само може бути невалідним! Мова також вимагає, щоб метод `return()` повертав об'єкт, а інакше - викидав TypeError. Якщо метод `return()` викидає помилку, то ця помилка поширюється до викликача. Проте якщо метод `return()` викликається, тому що викликач зіткнувся з помилкою у своєму власному коді, то така помилка замінює помилку, викинуту методом `return()`.

Reviewer: ```suggestion
Метод `return()` так само може бути невалідним! Мова також вимагає, щоб метод `return()` повертав об'єкт, а інакше викидав TypeError. Якщо метод `return()` викидає помилку, то ця помилка поширюється до викликача. Проте якщо метод `return()` викликається, тому що викликач зіткнувся з помилкою у своєму власному коді, то така помилка замінює помилку, викинуту методом `return()`.
```

Зазвичай викликач реалізує обробку помилок ось так:

```js
try {
  for (const value of iterable) {
    // ...
  }
} catch (e) {
  // Обробити помилку
}
```

`catch` зможе перехопити помилки, викинуті, коли `iterable` не є валідним ітерованим об'єктом, коли `next()` викидає помилку, коли `return()` викидає помилку (якщо цикл `for` завершується передчасно) або коли тіло циклу `for` викидає помилку.

Більшість ітераторів реалізовані за допомогою функцій-генераторів, тож ми продемонструємо, як функції-генератори зазвичай обробляють помилки:

```js
function* gen() {
  try {
    yield doSomething();
    yield doSomethingElse();
  } finally {
    cleanup();
  }
}
```

> Відсутність `catch` змушує помилки, викинуті `doSomething()` або `doSomethingElse()`, поширюватися до викликача `gen`. Якщо ці помилки перехоплюються всередині функції-генератора (що так само вітається), то така функція-генератор може вирішити продовжити видавати значення або завершитись передчасно. Проте блок `finally` необхідний для генераторів, що зберігають ресурси відкритими. Блок `finally` гарантовано запуститься, або коли востаннє викличеться `next()`, або коли викличеться `return()`.

Reviewer: ```suggestion
Відсутність `catch` змушує помилки, викинуті `doSomething()` або `doSomethingElse()`, поширюватися до викликача `gen`. Якщо ці помилки перехоплюються всередині функції-генератора (що так само вітається), то така функція-генератор може вирішити продовжити видавати значення або завершитися передчасно. Проте блок `finally` необхідний для генераторів, що зберігають ресурси відкритими. Блок `finally` гарантовано запуститься, коли або востаннє викличеться `next()`, або викличеться `return()`.
```

### Переспрямування помилок

> Частина вбудованих записів синтаксису загортають ітератор в інший ітератор. Серед них ітератор, що створюється {{jsxref("Iterator.from()")}}, [ітераторні помічники](/uk/docs/Web/JavaScript/Reference/Global_Objects/Iterator#iteratorni-pomichnyky) (`map()`, `filter()`, `take()`, `drop()` і `flatMap()`), [`yield*`](/uk/docs/Web/JavaScript/Reference/Operators/yield*) і прихована обгортка, коли щодо синхронного ітератора використовується асинхронна ітерація (`for await...of`, `Array.fromAsync`). Загорнутий ітератор тоді відповідає за переспрямування помилок між внутрішнім ітератором і викликачем.

Reviewer: ```suggestion
Частина вбудованих записів синтаксису загортає ітератор в інший ітератор. Серед них ітератор, що створюється методом {{jsxref("Iterator.from()")}}, [ітераторні помічники](/uk/docs/Web/JavaScript/Reference/Global_Objects/Iterator#iteratorni-pomichnyky) (`map()`, `filter()`, `take()`, `drop()` і `flatMap()`), [`yield*`](/uk/docs/Web/JavaScript/Reference/Operators/yield*) і прихована обгортка, коли щодо синхронного ітератора використовується асинхронна ітерація (`for await...of`, `Array.fromAsync`). Загорнутий ітератор тоді відповідає за переспрямування помилок між внутрішнім ітератором і викликачем.
```

Дієслово-присудок буде в однині й тоді, коли за підмет правлять збірні слова — більшість, меншість, частина, ряд, кілька, багато, чимало, група, низка, решта, половина та інші, поєднані з залежними від них іменниками (прикметниками) в родовому відмінку: «Більшість присутніх пристала на цю пропозицію»; «Кілька відер води бухнуло на вогонь» (М. Коцюбинський).

> - Усі ітератори-обгортки безпосередньо переспрямовують метод `next()` внутрішнього ітератора, в тому числі повернені значення та викинуті помилки.

Reviewer: ```suggestion
- Усі ітератори-обгортки безпосередньо переспрямовують метод `next()` внутрішнього ітератора, включно з поверненими значеннями та викинутими помилками.
```

> - Ітератори-обгортки, загалом, безпосередньо переспрямовують метод `return()` внутрішнього ітератора. Якщо метод `return()` на внутрішньому ітераторі не існує, то зовнішній повертає натомість `{ done: true, value: undefined }`. У разі ітераторних помічників: якщо метод `next()` ітераторного помічника ще не був викликаний, то після спроби викликати `return()` на внутрішньому ітераторі поточний ітератор завжди повертає `{ done: true, value: undefined }`. Ця поведінка узгоджена з функціями-генераторами, в яких плин виконання ще не дійшов до виразу `yield*`.

Reviewer: ```suggestion
- Ітератори-обгортки загалом безпосередньо переспрямовують метод `return()` внутрішнього ітератора. Якщо метод `return()` на внутрішньому ітераторі не існує, то зовнішній повертає натомість `{ done: true, value: undefined }`. У разі ітераторних помічників: якщо метод `next()` ітераторного помічника ще не був викликаний, то після спроби викликати `return()` на внутрішньому ітераторі поточний ітератор завжди повертає `{ done: true, value: undefined }`. Ця поведінка узгоджена з функціями-генераторами, в яких плин виконання ще не дійшов до виразу `yield*`.
```

> - `yield*` – це єдиний вбудований запис синтаксису, що переспрямовує метод `throw()` внутрішнього ітератора. Про те, як [`yield*`](/uk/docs/Web/JavaScript/Reference/Operators/yield*) переспрямовує методи `return()` і `throw()`, читайте в його власній довідці.

Reviewer: довідці чи довіднику?

Author: Мабуть, довідці, бо мова ж про одну сторіночку, а не цілу книгу
